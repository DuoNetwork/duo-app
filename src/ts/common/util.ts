import { IContractPrice } from '@finbook/duo-contract-wrapper';
import { IPriceStatus, IStatus } from '@finbook/duo-market-data';
import * as d3 from 'd3';
import moment, { DurationInputArg2 } from 'moment';
import { ISourceData } from './types';

class Util {
	public convertUpdateTime(timestamp: number): string {
		const diff = this.getUTCNowTimestamp() - timestamp;
		if (diff < 60000) return 'Just Now';
		else if (diff < 3600000) return Math.floor(diff / 60000) + ' Minutes Ago';
		else if (diff < 86400000) return Math.floor(diff / 3600000) + ' Hours Ago';
		else if (diff < 2592000000) return Math.floor(diff / 86400000) + ' Days Ago';
		else return 'Long Time Ago';
	}

	public formatTime(e: number) {
		return moment(e).format('YYYY-MM-DD HH:mm:ss');
	}

	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public getDates(length: number, step: number, stepSize: DurationInputArg2, format: string) {
		const dates: string[] = [];
		const date = moment.utc();
		for (let i = 0; i < length; i++) {
			dates.push(date.format(format));
			date.subtract(step, stepSize);
		}
		dates.sort((a, b) => a.localeCompare(b));

		return dates;
	}

	public round(num: number) {
		return +(Math.floor((num + 'e+8') as any) + 'e-8');
	}

	public formatBalance(num: number) {
		if (Math.abs(num) < 1e-8) return '0.000';
		return d3
			.format(Math.abs(num) > 1 ? ',.4s' : ',.4n')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatNumber(num: number) {
		if (Math.abs(num) < 1e-8) return '0.000';
		if (Math.abs(num) < 1) return d3.format(',.4n')(num);
		if (Math.abs(num) < 100000) return d3.format(',.2f')(num);
		return d3
			.format(',.4s')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public getLastPriceFromStatus(status: IStatus[]): ISourceData<IContractPrice> {
		// const bitfinex: ICustodianPrice = {
		// 	price: 0,
		// 	timestamp: 0
		// };
		const kraken: IContractPrice = {
			price: 0,
			timestamp: 0
		};
		const gemini: IContractPrice = {
			price: 0,
			timestamp: 0
		};
		const gdax: IContractPrice = {
			price: 0,
			timestamp: 0
		};
		status.forEach(s => {
			// if (s.process === 'PRICE_AWS_PUBLIC_BITFINEX') {
			// 	bitfinex.price = (s as IPriceStatus).price;
			// 	bitfinex.timestamp = (s as IPriceStatus).timestamp;
			// } else
			if (s.process === 'TRADE_AWS_PUBLIC_GEMINI') {
				gemini.price = (s as IPriceStatus).price;
				gemini.timestamp = (s as IPriceStatus).timestamp;
			} else if (s.process === 'TRADE_AWS_PUBLIC_KRAKEN') {
				kraken.price = (s as IPriceStatus).price;
				kraken.timestamp = (s as IPriceStatus).timestamp;
			} else if (s.process === 'TRADE_AWS_PUBLIC_GDAX') {
				gdax.price = (s as IPriceStatus).price;
				gdax.timestamp = (s as IPriceStatus).timestamp;
			}
		});

		return {
			// bitfinex,
			kraken,
			gemini,
			gdax
		};
	}
}

const util = new Util();
export default util;
