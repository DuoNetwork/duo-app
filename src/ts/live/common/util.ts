import * as d3 from 'd3';
import moment, { DurationInputArg2 } from 'moment';
import * as CST from './constants';
import { ICustodianPrice, IPriceStatus, ISourceData, IStatus } from './types';

class Util {
	public convertUpdateTime(timestamp: number): string {
		const diff = this.getUTCNowTimestamp() - timestamp;
		if (diff < 60000) return 'Just Now';
		else if (diff < 3600000) return Math.floor(diff / 60000) + ' Minutes Ago';
		else if (diff < 86400000) return Math.floor(diff / 3600000) + ' Hours Ago';
		else if (diff < 2592000000) return Math.floor(diff / 86400000) + ' Days Ago';
		else return 'Long Time Ago';
	}

	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public calculateNav(
		price: number,
		time: number,
		resetPrice: number,
		resetTime: number,
		alpha: number,
		beta: number,
		period: number,
		coupon: number
	) {
		const navParent = (price / resetPrice / beta) * (1 + alpha);

		const navA = 1 + Math.floor((time - resetTime) / 1000 / period) * coupon;
		const navAAdj = navA * alpha;
		if (navParent <= navAAdj) return [navParent / alpha, 0];
		else return [navA, navParent - navAAdj];
	}

	public getDates(length: number, step: number, stepSize: DurationInputArg2, format: string) {
		const dates: string[] = [];
		const date = moment.utc();
		for (let i = 0; i < length; i++) {
			dates.push(date.format(format));
			date.subtract(step, stepSize);
		}
		dates.sort((a, b) => a.localeCompare(b));

		return dates;
	}

	public round(num: number) {
		return +(Math.floor((num + 'e+8') as any) + 'e-8');
	}

	public range(start: number, end: number) {
		const subArray = Array.apply(null, { length: end }).map(Number.call, Number);
		return subArray.slice(start, end);
	}

	public formatBalance(num: number) {
		if (Math.abs(num) < 1e-8) return '0.000';
		return d3
			.format(Math.abs(num) > 1 ? ',.4s' : ',.4n')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatNumber(num: number) {
		if (Math.abs(num) < 1e-8) return '0.000';
		if (Math.abs(num) < 1) return d3.format(',.4n')(num);
		if (Math.abs(num) < 100000) return d3.format(',.2f')(num);
		return d3
			.format(',.4s')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public getCurrentPath(location: any) {
		return location.pathname === '/' ? location.pathname + 'beethoven' : location.pathname;
	}

	public getLastPriceFromStatus(status: IStatus[]): ISourceData<ICustodianPrice> {
		// const bitfinex: ICustodianPrice = {
		// 	address: CST.DUMMY_ADDR,
		// 	price: 0,
		// 	timestamp: 0
		// };
		const kraken: ICustodianPrice = {
			address: CST.DUMMY_ADDR,
			price: 0,
			timestamp: 0
		};
		const gemini: ICustodianPrice = {
			address: CST.DUMMY_ADDR,
			price: 0,
			timestamp: 0
		};
		const gdax: ICustodianPrice = {
			address: CST.DUMMY_ADDR,
			price: 0,
			timestamp: 0
		};
		status.forEach(s => {
			// if (s.process === 'PRICE_AWS_PUBLIC_BITFINEX') {
			// 	bitfinex.price = (s as IPriceStatus).price;
			// 	bitfinex.timestamp = (s as IPriceStatus).timestamp;
			// } else
			if (s.process === 'TRADE_AWS_PUBLIC_GEMINI') {
				gemini.price = (s as IPriceStatus).price;
				gemini.timestamp = (s as IPriceStatus).timestamp;
			} else if (s.process === 'TRADE_AWS_PUBLIC_KRAKEN') {
				kraken.price = (s as IPriceStatus).price;
				kraken.timestamp = (s as IPriceStatus).timestamp;
			} else if (s.process === 'TRADE_AWS_PUBLIC_GDAX') {
				gdax.price = (s as IPriceStatus).price;
				gdax.timestamp = (s as IPriceStatus).timestamp;
			}
		});

		return {
			// bitfinex,
			kraken,
			gemini,
			gdax
		};
	}
}

const util = new Util();
export default util;
