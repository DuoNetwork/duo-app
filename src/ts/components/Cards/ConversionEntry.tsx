//import { Table, Tooltip } from 'antd';
import * as d3 from 'd3';
import * as React from 'react';
import * as CST from 'ts/common/constants';
import util from 'ts/common/util';

interface IProps {
	timestamp: number;
	type: string;
	eth: number;
	tokenACode: string;
	tokenA: number;
	tokenBCode: string;
	tokenB: number;
	pending?: boolean;
	reverted?: boolean;
	locale: string;
}

export default class ConversionEntry extends React.Component<IProps> {
	public render() {
		const {
			timestamp,
			type,
			eth,
			tokenA,
			tokenB,
			tokenACode,
			tokenBCode,
			pending,
			reverted,
			locale
		} = this.props;
		const tokenString =
			d3.format(',.2f')(tokenA) +
			' ' +
			tokenACode +
			'/' +
			d3.format(',.2f')(tokenB) +
			' ' +
			tokenBCode;
		return (
			<ul>
				<li>
					<span className="title">{util.formatTime(timestamp)}</span>
					<span className="content">
						{pending
							? CST.TH_PENDING[locale]
							: reverted
							? CST.TH_REVERTED[locale]
							: CST.TH_MINED[locale]}
					</span>
				</li>
				<li>
					<span className="title">
						{type === 'Create' ? CST.TH_CREATE[locale] : CST.TH_REDEEM[locale]}
					</span>
					<span className="content">
						{type === 'Create'
							? d3.format(',.2f')(eth) + ' ETH => ' + tokenString
							: tokenString + ' => ' + d3.format(',.2f')(eth) + ' ETH'}
					</span>
				</li>
			</ul>
		);
	}
}
