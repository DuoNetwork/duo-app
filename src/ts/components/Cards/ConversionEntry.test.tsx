import { shallow } from 'enzyme';
import * as React from 'react';
import util from 'ts/common/util';
import ConversionEntry from './ConversionEntry';

describe('ConversionEntry Test', () => {
	describe('Test Snapshot', () => {
		const timestamp = 123;
		const type = 'test';
		const eth = 123;
		const tokenACode = 'test';
		const tokenA = 123;
		const tokenBCode = 'test';
		const tokenB = 123;
		const pending = true;
		const reverted = true;
		const locale = 'test';
		it('Test Snapshot', () => {
			util.formatTime = jest.fn(() => '1970-01-01 12:00:00');
			const wrapper = shallow(
				<ConversionEntry
					timestamp={timestamp}
					type={type}
					eth={eth}
					tokenACode={tokenACode}
					tokenA={tokenA}
					tokenBCode={tokenBCode}
					tokenB={tokenB}
					pending={pending}
					reverted={reverted}
					locale={locale}
				/>
			);
			expect(wrapper).toMatchSnapshot();
		});
	});
});
