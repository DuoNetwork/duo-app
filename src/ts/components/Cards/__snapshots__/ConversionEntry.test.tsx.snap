// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversionEntry Test Test Snapshot Test Snapshot 1`] = `
<ul>
  <li>
    <span
      className="title"
    >
      1970-01-19 03:44:03
    </span>
    <span
      className="content"
    >
      Pending
    </span>
  </li>
  <li>
    <span
      className="title"
    >
      Redeem
    </span>
    <span
      className="content"
    >
      1.00 Token A/B =&gt; 1.00 ETH
    </span>
  </li>
</ul>
`;
